clients.go

```diff
@@ -26,24 +34,57 @@ func (client *KustoIngestionClient) SendAsync(msg PubsubMsg) error {
 		log.Fatalf("Failed to set function key env var")
 	}
 
-	requestUri := targetUri + "?code=" + functionKey
+	INGESTIONSERVICEBATCHSIZE := os.Getenv("INGESTIONSERVICEBATCHSIZE")
+	batchSize, err := strconv.Atoi(INGESTIONSERVICEBATCHSIZE)
+	if INGESTIONSERVICEBATCHSIZE == "" || err != nil {
+		log.Fatalf("Failed to set INGESTIONSERVICEBATCHSIZE %v", err)
+	}
+
+	requestUri := targetUri + "?code=" + functionKey;
+
+	pubSubMsgs := make([]PubsubMsg, 0, batchSize)
+
+	return KustoIngestionClient { 
+		requestUri:requestUri,
+		pubSubMsgs: pubSubMsgs,
+		batchSize: batchSize,
+	}
+}
+
+func (client *KustoIngestionClient) SendAsync(msg PubsubMsg) error {
+	client.mutex.Lock()
+	defer client.mutex.Unlock()
+
+	client.pubSubMsgs = append(client.pubSubMsgs, msg);
+	client.counter += 1
+
+	if client.counter == client.batchSize {
+		msgs := make([]PubsubMsg, client.batchSize, client.batchSize)
+		copy(msgs, client.pubSubMsgs)
 
-	rawMsgs := RawMsgs{
-		Records: []PubsubMsg{msg},
+		client.counter = 0
+		client.pubSubMsgs = make([]PubsubMsg, 0, client.batchSize)
+		return client.SendAsyncBatch(msgs)
 	}
 
+	return nil
+}
+
+func (client *KustoIngestionClient) SendAsyncBatch(msgs []PubsubMsg) error {
+	rawMsgs := RawMsgs{Records: msgs}
+
 	buf, err := json.Marshal(rawMsgs)
 	if err != nil {
 		log.Fatalf(err.Error())
 	}
 
 	var resp *http.Response
 	for i := 0; i < 3; i++ {
```

Comment:
Does this mean we retry a max of 3 times?

Reply:
Yeah arbitrary count